### PR TITLE
Feature: Implement Instance snapshots

### PR DESCRIPTION
**Pull request by nesitor**
Feature: Upload VM snapshots using Aleph STORE message.
Feature: Restore existing snapshots as STORE Aleph message.

### CODE REVIEW

File: vm_supervisor/snapshots.py [hunk 1]
```diff
@@ -29,9 +34,13 @@ def __init__(self, path: Path, algorithm: SnapshotCompressionAlgorithm):
...
-    async def upload(self) -> ItemHash:
-        # TODO: Upload snapshots to Aleph Network
-        pass
+    async def upload(self, vm_hash: ItemHash) -> ItemHash:
+        account = get_fallback_account()
+        async with AuthenticatedAlephClient(account=account, api_server="https://official.aleph.cloud") as client:
+            message, status = await client.create_store(
+                file_path=self.path, storage_engine=StorageEnum.ipfs, sync=True, ref=f"snapshot_{vm_hash}"
+            )
+            assert status == MessageStatus.PROCESSED
```
Comment (odesenfans) [inline vm_supervisor/snapshots.py:43]: I think it's important here to delete old snapshots from the network as well. A call to `client.forget()`would make a lot of sense when we delete the previous snapshot. We'll implement some depth to this (keep the last N snapshots) later.

Comment (nesitor) [reply vm_supervisor/snapshots.py:54]: Good point, thanks, I just Implemented it.

File: vm_supervisor/ipfs.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+async def ipfs_upload_file(path: Path) -> str:
+    """Push a file to the IPFS service."""
+    # Use local ipfs node by default. TODO: Add a general setting to allow to customize it
+    client = aioipfs.AsyncIPFS(host="localhost", port=5001)
+    ipfs_hash = None
+    async for added_file in client.add(str(path), recursive=True):
```
Comment (hoh) [inline vm_supervisor/ipfs.py:21]: Why is this recursive if the goal is to upload a single file ?

File: vm_supervisor/ipfs.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            ref=ref,
+        )
+        assert status != MessageStatus.REJECTED
+        return message.item_hash
+
+
+async def send_forget_ipfs_message(
+    item_hash: ItemHash, reason: Optional[str] = ""
+) -> None:
+    pkey = get_fallback_private_key()
```
Comment (hoh) [inline vm_supervisor/ipfs.py:59]: This function should never be used in production: the creation of the key should be explicit and during installation. A key that is missing should cause an error.

File: vm_supervisor/ipfs.py (hunk 2 above)
Comment (nesitor) [reply vm_supervisor/ipfs.py:21]: Because it accepts different files, and it return different results.

File: vm_supervisor/ipfs.py (hunk 3 above)
Comment (nesitor) [reply vm_supervisor/ipfs.py:59]: How I can get it?

File: vm_supervisor/ipfs.py (hunk 2 above)
Comment (hoh) [reply vm_supervisor/ipfs.py:21]: In practice I only see it used in one place, `snapshot_hash = await ipfs_upload_file(self.path)`, with a single file.

Comment (nesitor) [reply vm_supervisor/ipfs.py:21]: Yes, I mean that the `client.add` method call, it accepts different files, and return different results.

**Comment by hoh** [comment]
I am closing this PR since we are facing performance issues in this implementation and want to evaluate the use of Qcow2 as a replacement.

**Pull request closed without merging**